[rd03d] Add Ai-Thinker RD-03D mmWave radar component

jasstrong opened the pull request:
# What does this implement/fix?

Support for the RD-03D 24GHz millimeter-wave radar module with multi-target trajectory tracking. Tracks up to 3 simultaneous targets with x, y, speed, distance, angle, and resolution sensors.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5839

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
  uart:
    id: uart_rd03d
    tx_pin: GPIO17
    rx_pin: GPIO16
    baud_rate: 256000

  rd03d:
    id: rd03d_radar
    uart_id: uart_rd03d

  binary_sensor:
    - platform: rd03d
      rd03d_id: rd03d_radar
      target:
        name: Presence
      target_1:
        name: Target-1 Presence

  sensor:
    - platform: rd03d
      rd03d_id: rd03d_radar
      target_count:
        name: Target Count
      target_1:
        x:
          name: Target-1 X
        y:
          name: Target-1 Y
        speed:
          name: Target-1 Speed
        distance:
          name: Target-1 Distance
        angle:
          name: Target-1 Angle


```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
Add Ai-Thinker RD-03D mmWave radar component

The RD-03D is a 24GHz millimeter-wave radar module that can simultaneously
track up to 3 moving targets, providing real-time position (X/Y coordinates),
speed, and distance information for each target.

Features:
- Multi-target tracking (up to 3 targets)
- Per-target sensors: X, Y, speed, distance, angle, resolution
- Per-target and overall presence binary sensors
- Optional tracking mode configuration (single/multi target)

swoboda1337 commented on tests/components/rd03d/.gitignore:
remove this file

jasstrong replied:
oops, sorry about that.

swoboda1337 commented on tests/components/rd03d/test.esp32-idf.yaml:
Take a look at other uart tests like https://github.com/esphome/esphome/blob/dev/tests/components/ld2450/test.esp32-idf.yaml

Should remove esphome, esp32, logger etc and include a common uart file.

jasstrong replied:
The issue is the common uart files use 9600 baud, but RD-03D needs 256000.  I'll take a look at how LD2450 does this because it's similarly high rate

swoboda1337 replied:
There are a few different versions of the uart files: https://github.com/esphome/esphome/tree/dev/tests/test_build_components/common/uart_115200

jasstrong replied:
ah, can't do that-  the LD2450 has configurable baud rate and the RD-03D does not, so I can't use a standard UART package.  I'll have to create one.

swoboda1337 replied:
You are also not validating the baud rate so it doesnt really matter these are not runtime tests.

swoboda1337 replied:
Although maybe you should validate the baud rate then?

jasstrong replied:
the RD-03D has firmware build time selectable baud rate (though its firmware is not open source) so I didn't validate the baud rate in case someone has a module that's configured differently.

Given that we don't run against real hardware I guess you're right, we might as well just use the 9600 rate-  though that won't test for a platform that isn't capable of 256000 baud.

jasstrong replied:
That's not a compile-time test anyway so I guess the test won't catch it either.

swoboda1337 commented on esphome/components/rd03d/rd03d.cpp:
```diff
@@ -0,0 +1,204 @@
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace rd03d {
```
```suggestion
namespace esphome::rd03d {
```
Should use modern namespace syntax

jasstrong replied:
that'll teach me to crib from older sensors ;-)

swoboda1337 commented on esphome/components/rd03d/__init__.py:
```diff
@@ -0,0 +1,52 @@
+
+CODEOWNERS = ["@jasstrong"]
+DEPENDENCIES = ["uart"]
+AUTO_LOAD = ["sensor", "binary_sensor"]
```
Should remove this and add USE_SENSOR ifdefs

jasstrong replied:
done

swoboda1337 commented on esphome/components/rd03d/rd03d.cpp:
```diff
@@ -0,0 +1,204 @@
+
+void RD03DComponent::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up RD-03D...");
+  this->set_timeout(100, [this]() { this->apply_config_(); });
```
Magic number, could use a constexpr and a comment

jasstrong replied:
yeah, that's a good idea for sure.

swoboda1337 commented on esphome/components/rd03d/rd03d.h:
```diff
@@ -0,0 +1,107 @@
+
+// Decode coordinate/speed value from RD-03D format
+// Per datasheet: MSB=1 means positive, MSB=0 means negative
+static inline int16_t decode_value(uint8_t low_byte, uint8_t high_byte) {
```
Instead of inline could be constexpr

swoboda1337 commented on esphome/components/rd03d/sensor.py:
```diff
@@ -0,0 +1,107 @@
+
+UNIT_MILLIMETER_PER_SECOND = "mm/s"
+
+TargetSensor = rd03d_ns.struct("TargetSensor")
```
Never used?

jasstrong replied:
ah, that's dead code.

swoboda1337 commented on esphome/components/rd03d/sensor.py:
```diff
@@ -0,0 +1,107 @@
+
+MAX_TARGETS = 3
+
+UNIT_MILLIMETER_PER_SECOND = "mm/s"
```
We cant use an existing unit?

swoboda1337 commented on tests/components/rd03d/common.yaml:
```diff
@@ -0,0 +1,61 @@
+rd03d:
+  id: rd03d_radar
+  # tracking_mode: multi  # optional, only sends command if set
+  # max_distance_gate: 8  # experimental
```
Commented out?

jasstrong replied:
max_distance_gate should go away for sure;  tracking_mode is supported, but generally not necessary (multi is the default for most firmware builds)

jasstrong replied:
got rid of them for cleanliness

swoboda1337 commented on esphome/components/rd03d/rd03d.cpp:
```diff
@@ -0,0 +1,204 @@
+
+  // Calculate and publish distance (mm)
+  if (target.distance != nullptr) {
+    float distance = std::sqrt(static_cast<float>(x) * x + static_cast<float>(y) * y);
```
Can use std::hypot

Commit:
Address review feedback for RD-03D component

- Use modern C++ namespace syntax (namespace esphome::rd03d)
- Replace inline with constexpr for decode_value()
- Extract setup timeout as named constexpr SETUP_TIMEOUT_MS
- Use std::hypot() for distance calculation
- Add USE_SENSOR/USE_BINARY_SENSOR ifdefs
- Remove AUTO_LOAD, let platforms be explicitly configured
- Remove commented code from test yaml

Commit:
Remove unused TargetSensor declaration from sensor.py

swoboda1337 committed:
Clean up

swoboda1337 commented on esphome/components/rd03d/rd03d.cpp:
```diff
@@ -0,0 +1,243 @@
+
+void RD03DComponent::process_frame_() {
+  // Apply throttle if configured
+  if (this->throttle_ > 0) {
+    uint32_t now = millis();
+    if (now - this->last_publish_time_ < this->throttle_) {
+      return;
+    }
+    this->last_publish_time_ = now;
+  }
```
Do we need this? Why not just use a sensor filter?

jasstrong replied:
You would have to set a sensor filter on all of the fifteen or so sensors this device exposes and it was just a lot to deal with.  I can drop this if you prefer, but it was a lot of yaml spam.

jasstrong replied:
This also saves a lot of CPU time on constrained devices like original esp8266, which can easily fall behind  trying to handle a bajillion sensor filter invocations a second.

swoboda1337 replied:
How fast can it update? How many times a minute?

jasstrong replied:
It generates updates almost as fast as the serial port supports-  every few milliseconds.

swoboda1337 replied:
Ok

swoboda1337 replied:
Ok

jasstrong replied:
```
[15:21:01.163][D][rd03d:063]: Valid frame received
[15:21:01.176][D][sensor:135]: 'Target-1 X': Sending state 0.00000 mm with 0 decimals of accuracy
[15:21:01.176][D][sensor:135]: 'Target-1 Y': Sending state 0.00000 mm with 0 decimals of accuracy
[15:21:01.176][D][sensor:135]: 'Target-1 Speed': Sending state 0.00000 mm/s with 0 decimals of accuracy
[15:21:01.176][D][sensor:135]: 'Target-1 Distance': Sending state 0.00000 mm with 0 decimals of accuracy
[15:21:01.176][D][sensor:135]: 'Target-1 Angle': Sending state 0.00000 ° with 1 decimals of accuracy
[15:21:01.176][D][sensor:135]: 'Target Count': Sending state 0.00000  with 0 decimals of accuracy
[15:21:01.258][D][rd03d:063]: Valid frame received
```

an update about every 95 milliseconds-  I only had one of the three targets configured here, so there could have been 3x this many updates.
